modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryEventBuffer.java

```diff
@@ -322,6 +332,21 @@ private Object process0(long cntr, CacheContinuousQueryEntry entry, boolean back
             endCntr = startCntr + BUF_SIZE - 1;
         }
 
+        /**
+         * @param updateCntr Acknowledged counter.
+         */
+        synchronized void cleanupEntries(Long updateCntr) {
+            if (entries == null)
+                return;
+
+            for (int i = 0; i < entries.length; i++) {
```

Comment:
I suppose we do not need to visit each item in `entries`. Can we remember last cleaned index for a batch and start a cleanup from it? Also we can stop the loop as long as first entry with greater update counter is met.